datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -414,7 +415,18 @@ fn get_valid_types(
             _ => Ok(vec![vec![]]),
         }
     }
+
     fn array(array_type: &DataType) -> Option<DataType> {
+        match array_type {
+            DataType::List(_) => Some(array_type.clone()),
+            DataType::LargeList(field) | DataType::FixedSizeList(field, _) => {
+                Some(DataType::List(Arc::clone(field)))
+            }
+            _ => None,
+        }
+    }
+
+    fn recursive_array(array_type: &DataType) -> Option<DataType> {
```

Comment:
Can we extend the existing array function for nested array instead of creating another signature for nested array

Reply:
I don't know how to do this, please advise!
But this function should go away with https://github.com/apache/datafusion/issues/13757.

Reply:
> But this function should go away with #13757.

I don't understand -- if the goal is to remove recursive flattening, should we be adding new code to support it 🤔

Reply:
the pre-existing `array` signature implied recursively array-infication (replacing FixedLengthList with List, recursively), didn't imply flattening.

the recursive type normalization matters for flatten only, cause it (currently) operates recursively and otherwise would need to gain code to handle FixedLengthList inputs

the recursive array-ification was useless for other array functions and was made non-recursive.
to compensate for this change, new RecursiveArray signature was added for flatten case.